### PR TITLE
Added "Specialized Workload" - HPC contents to APRL AB#31655

### PR DESCRIPTION
**Pull request by lapate**
# Overview/Summary

To continuously improve our WAF Reliability offering we are initializing analysis and better tracking of recommendations (Azure Services & WAF) through their lifecycle. That lifecycle consists of the following 5 stages some of which can happen in parallel:

Create guidance and resource graph query in APRL
Push resource graph query to Advisor Reliability Workbook
Push recommendation into CX Observe
Push recommendation into Azure Advisor
Publish guidance & query to public docs (Ex: VM Reliability)

This PR is for - **Create guidance and resource graph query in APRL**

## Related Issues/Work Items

AB #31655

## This PR fixes/adds/changes/removes

1. Initial update to the contents for HPC
2. Updated/Removed not relevant Azure Graph Queries

### Breaking Changes

1. *Replace me*
2. *Replace me*

## As part of this Pull Request I have

- [x] Read the [Contribution Guide](https://azure.github.io/Azure-Proactive-Resiliency-Library/contributing) and ensured this PR is compliant with the guide
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Azure-Proactive-Resiliency-Library/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/Azure-Proactive-Resiliency-Library/issues) or ADO Work Items (Internal Only)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Azure-Proactive-Resiliency-Library/tree/main)
- [x] Ensured PR tests are passing
- [x] Performed testing and provided evidence (e.g. screenshot of output) for any changes associated to ARG queries and/or scripts
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Docs etc.)

### CODE REVIEW

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+### HPC-3 - Ensure Azure File shares are active
+
+**Category: Application Resilience/Availability**
+
+**Impact: High**
+
+**Recommendation/Guidance**
+
+Currently in all HPC Pack ARM templates we create the cluster share on one of the head node which is not high available as if that head node is down, the share will not be accessible to the HPC Service running on other head node. Basically it will not impact running jobs and managing the nodes.
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:101]: Please reword for clarity as it is a bit difficult to understand what is being stated and how it relates to the recommendation of 'Ensure Azure File shares are active'

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+With Azure Files, these file shares can be moved to Azure Files shares with SMB permissions to make them high available.
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:103]:
```suggestion
With Azure Files, the following file shares can be moved to Azure Files shares with SMB permissions to make them highly available:
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+{{< code lang="sql" file="code/hpc-3/hpc-3.kql" >}} {{< /code >}}
+
+{{< /collapse >}}
+
+<br><br>
+
+### HPC-4 - Create an Azure Batch pool across Availability Zones
+
+**Category: Availability/Identity**
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:129]:
```suggestion
**Category: Availability**
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+{{< code lang="sql" file="code/hpc-4/hpc-4.kql" >}} {{< /code >}}
+
+{{< /collapse >}}
+
+<br><br>
+
+### HPC-5 - Automatically grow and shrink HPC Pack cluster resources
+
+**Category: Availability**
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:154]:
```suggestion
**Category: System Efficiency**
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+| :------------------------------------------------ | :---------------------------------------------------------------------: | :------:        | :------:          | :------:          |
+| [HPC-1 Ensure that Storage Account is redundant](#hpc-1---ensure-that-storage-account-is-redundant)    | High | Resiliency/Monitoring |  Preview  |        Yes         |
+| [HPC-2 Monitor Azure Batch Pool quota for cross-region disaster recovery and business continuity](#hpc-2---monitor-batch-account-quota)  | Medium |  Resiliency/Monitoring | Preview |       Yes        |
+| [HPC-3 Ensure Azure File shares Are active](#hpc-3---ensure-azure-file-shares-are-active)  | High |  Resiliency/Monitoring | Preview |       Yes        |
+| [HPC-4 Use Availability Zone for batch pool](#hpc-4---create-an-azure-batch-pool-across-availability-zones)  | Medium |  Availability | Preview |       No        |
+| [HPC-5 Automatically grow and shrink HPC Pack cluster resources](#hpc-5---automatically-grow-and-shrink-hpc-pack-cluster-resources)  | Medium |  Availability | Preview |       No        |
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:21]:
```suggestion
| [HPC-5 Automatically grow and shrink HPC Pack cluster resources](#hpc-5---automatically-grow-and-shrink-hpc-pack-cluster-resources)  | Medium |  System Efficiency | Preview |       No        |
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+### HPC-5 - Automatically grow and shrink HPC Pack cluster resources
+
+**Category: Availability**
+
+**Impact: Medium**
+
+**Recommendation/Guidance**
+
+By deploying Azure "burst" nodes (both Windows and Linux) in your HPC Pack cluster or creating your HPC Pack cluster in Azure, you can automatically grow or shrink the cluster's resources such as nodes or cores according to the workload on the cluster. Scaling the cluster resources in this way allows you to use your Azure resources more efficiently.
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:160]: Can you update the wording to reflect how this impacts reliability directly.

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+{{< collapse title="Show/Hide Query/Script" >}}
+
+{{< code lang="sql" file="code/hpc-5/hpc-5.kql" >}} {{< /code >}}
+
+{{< /collapse >}}
+
+<br><br>
+
+### HPC-6 - HPC Pack - Dealing with database failure
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:176]: Like the storage account recommendation above, this recommendation seems to be generalized to creating highly available SQL database. Please remove unless there is anything specific that needs to be called out from an HPC perspective.

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+**Resources**
+
+- [Learn More](https://learn.microsoft.com/en-us/powershell/high-performance-computing/hpcpack-ha-cloud?view=hpc19-ps#dealing-with-database-failure)
+
+<br><br>
+
+
+
+### HPC-7 - HPC Pack - Dealing with Head node failure
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:196]:
```suggestion
### HPC-7 - HPC Pack - Dealing with head node failure
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+| [HPC-2 Monitor Azure Batch Pool quota for cross-region disaster recovery and business continuity](#hpc-2---monitor-batch-account-quota)  | Medium |  Resiliency/Monitoring | Preview |       Yes        |
+| [HPC-3 Ensure Azure File shares Are active](#hpc-3---ensure-azure-file-shares-are-active)  | High |  Resiliency/Monitoring | Preview |       Yes        |
+| [HPC-4 Use Availability Zone for batch pool](#hpc-4---create-an-azure-batch-pool-across-availability-zones)  | Medium |  Availability | Preview |       No        |
+| [HPC-5 Automatically grow and shrink HPC Pack cluster resources](#hpc-5---automatically-grow-and-shrink-hpc-pack-cluster-resources)  | Medium |  Availability | Preview |       No        |
+| [HPC-6 HPC Pack - Dealing with database failure](#hpc-6---hpc-pack---dealing-with-database-failure)  | Medium |  Resiliency | Preview |       No        |
+| [HPC-7 HPC Pack - Dealing with Head node failure](#hpc-7---hpc-pack---dealing-with-head-node-failure)  | Medium |  Resiliency | Preview |       No        |
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:23]:
```suggestion
| [HPC-7 HPC Pack - Dealing with head node failure](#hpc-7---hpc-pack---dealing-with-head-node-failure)  | Medium |  Resiliency | Preview |       No        |
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+### HPC-7 - HPC Pack - Dealing with Head node failure
+
+**Category: Resiliency**
+
+**Impact: Medium**
+
+**Recommendation/Guidance**
+
+Set up at least 2 head nodes in a cluster. With this configuration, any head node failure will result in moving the active HPC Service from this head node to another.
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:204]:
```suggestion
Establish a cluster with a minimum of two head nodes. In the event of a head node failure, the active HPC Service will be automatically transferred from the affected head node to another functioning one.
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+**Recommendation/Guidance**
+
+When HPC failed to connect to the Domain controller, admin and user will not be able to connect to the HPC Service thus not able to manage and submit jobs to the cluster. And new jobs will not be able started on the domain joined computer nodes as the NodeManager service failed to validate the job's credential. Thus you need consider below options:
+
+Having a high available domain controller deployed with your HPC Pack Cluster in Azure
+
+Using Azure AD Domain service. During cluster deployment, you could just join all your cluster nodes into this domain and you get the high available domain service from Azure.
+
+Using HPC Pack Azure AD integration solution without having the cluster nodes joining any domain. Thus as long as the HPC Service has connectivity to the Azure AD service.
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:226]:
```suggestion
- Having a high available domain controller deployed with your HPC Pack Cluster in Azure

- Using Azure AD Domain service. During cluster deployment, you could just join all your cluster nodes into this domain and you get the high available domain service from Azure.

- Using HPC Pack Azure AD integration solution without having the cluster nodes joining any domain. Thus as long as the HPC Service has connectivity to the Azure AD service.
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+Set up at least 2 head nodes in a cluster. With this configuration, any head node failure will result in moving the active HPC Service from this head node to another.
+
+**Resources**
+
+- [Learn More](https://learn.microsoft.com/en-us/powershell/high-performance-computing/hpcpack-ha-cloud?view=hpc19-ps#dealing-with-head-node-failure)
+
+<br><br>
+
+### HPC-8 - HPC Pack - Dealing with AD failure
```
Comment (oZakari) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:212]: Please adjust recommendation name to short snippet of what the recommendation is such as `Use HPC Pack Azure AD Integration or other highly available AD configuration`

File: docs/content/services/specialized-workloads/azure-hpc/_index.md (hunk 8 above)
Comment (oZakari) [reply docs/content/services/specialized-workloads/azure-hpc/_index.md:196]: Also, change to what the recommendation is versus just a generalized scenario could be `Set up at least 2 head nodes in a cluster`

**Review by oZakari — CHANGES_REQUESTED** [review]
Thanks for the contribution @lapate, along with the adjusting the review items associated to this review. Could you also please update the PR form with the relevant description and go through the [ARG query guidance](http://localhost:1313/Azure-Proactive-Resiliency-Library/contributing/#azure-resource-graph-arg-queries) to ensure all queries are in adherence. Once complete, please share screenshots of the results returned for all queries.

Thank you and please reach out if you have any questions!

**Comment by lapate** [comment]
@oZakari , Thanks for the review of the contents.  I have committed all the updates.

**Comment by oZakari** [comment]
Thanks @lapate! Could you update the PR form with the description of the PR as link any work items you're team may have assigned you.

image

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+### HPC-2 - Monitor batch account quota
+
+**Category: Resiliency/Monitoring**
+
+**Impact: Medium**
+
+**Recommendation/Guidance**
+
+For Cross-region disaster recovery and business continuity, Make sure ahead of time that the appropriate quotas are set for all user subscription Batch accounts, to allocate the required number of cores using the Batch account.
```
[the remark this answers is not in the document]
Comment (oZakari) [reply docs/content/services/specialized-workloads/azure-hpc/_index.md]: This looks to have been mistakenly resolved, but no changes appear to have been made to the recommendation guidance so unresolving.

**Review by oZakari — APPROVED** [review]
Thanks much @lapate!

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
++++
+
+The presented resiliency recommendations in this guidance include Azure High Performance Computing and associated resources and settings.
+
+## Summary of Recommendations
+
+{{< table style="table-striped" >}}
+|  Recommendation                                   |      Impact         |  Design Area         |  State            | ARG Query Available |
+| :------------------------------------------------ | :---------------------------------------------------------------------: | :------:        | :------:          | :------:          |
+| [HPC-1 Monitor Batch account quota](#hpc-1---monitor-batch-account-quota)  | Medium |  Resiliency/Monitoring | Preview |       Yes        |
```
Comment (tksh164) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:17]:
```suggestion
| [HPC-1 Monitor Batch account quota](#hpc-1---monitor-batch-account-quota)  | Medium |  Resiliency/Monitoring | Preview |       No        |
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+| [HPC-2 Ensure File shares that stores jobs metadata are accessible from all head nodes](#hpc-2---ensure-file-shares-that-stores-jobs-metadata-are-accessible-from-all-head-nodes)  | High |  Resiliency/Monitoring | Preview |       Yes        |
```
Comment (tksh164) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:18]:
```suggestion
| [HPC-2 Ensure File shares that stores jobs metadata are accessible from all head nodes](#hpc-2---ensure-file-shares-that-stores-jobs-metadata-are-accessible-from-all-head-nodes)  | High |  Resiliency/Monitoring | Preview |       No        |
```

File: docs/content/services/specialized-workloads/azure-hpc/_index.md [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+Currently in all HPC Pack ARM templates we create the cluster share on one of the head node which is not highly available. If that head node is down, the share will not be accessible to the HPC Service running on other head node.
+
+With Azure Files, the following file shares can be moved to Azure Files shares with SMB permissions to make them highly available:
+
+\\<HN3>\REMINST <br>
+\\<HN3>\HpcServiceRegistration <br>
+\\<HN3>\Runtime$ <br>
+\\<HN3>\TraceRepository <br>
+\\<HN3>\Diagnostics <br>
+\\<HN3>\CcpSpoolDir
```
Comment (tksh164) [inline docs/content/services/specialized-workloads/azure-hpc/_index.md:77]:
```suggestion
- `\\<HN3>\REMINST`
- `\\<HN3>\HpcServiceRegistration`
- `\\<HN3>\Runtime$`
- `\\<HN3>\TraceRepository`
- `\\<HN3>\Diagnostics`
- `\\<HN3>\CcpSpoolDir`
```

**Review by tksh164 — COMMENTED** [review]
@lapate Thank you for your contribution! Added suggestions, could you check those?

**Comment by tksh164** [comment]
@oZakari We also have to update the [services-abbreviations.csv](https://github.com/Azure/Azure-Proactive-Resiliency-Library/blob/main/services-abbreviations.csv). But cannot directly map HPC to the specific Azure resource type and the [CAF Abbreviation](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations). Any ideas?

**Comment by oZakari** [comment]
> @oZakari We also have to update the [services-abbreviations.csv](https://github.com/Azure/Azure-Proactive-Resiliency-Library/blob/main/services-abbreviations.csv). But cannot directly map HPC to the specific Azure resource type and the [CAF Abbreviation](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations). Any ideas?

Good question, I assumed since it is generalized to a workload that we wouldn't add it that services list. However, I don't have a lot of experience with HPC workloads so hoping @lapate can add clarity to determine if it makes sense to add something specific to that list for HPC.

**Comment by tksh164** [comment]
> > @oZakari We also have to update the [services-abbreviations.csv](https://github.com/Azure/Azure-Proactive-Resiliency-Library/blob/main/services-abbreviations.csv). But cannot directly map HPC to the specific Azure resource type and the [CAF Abbreviation](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations). Any ideas?
> 
> Good question, I assumed since it is generalized to a workload that we wouldn't add it that services list. However, I don't have a lot of experience with HPC workloads so hoping @lapate can add clarity to determine if it makes sense to add something specific to that list for HPC.

I just concerned the compatible with `wara-arg-runner.ps1` that use in WARAv3 because the script uses the `services-abbreviations.csv` in it. So, I quickly reviewed that script, it looks like the script will work well. (I have not actually tried it)

**Review by tksh164 — APPROVED** [review]
LGTM 👍

**Pull request merged**